### PR TITLE
RSDK-2322 - implement support for component dependencies in config

### PR DESCRIPTION
**Pull request by gvaradarajan**
This supports _implicit_ dependencies only; it does not make use of the 'depends_on' field (although this could be implemented in a later PR). I made this choice because it seems that we were headed in the direction of implicit rather than explicit dependencies in RDK anyway. The board is still always passed as a dependency to every component constructor.

Follow-up PRs (still under RSDK-2322) will implement dependency functions for the rest of the components for which such a thing is relevant.

### CODE REVIEW

File: src/common/registry.rs [hunk 1]
```diff
@@ -27,44 +27,112 @@ pub enum RegistryError {
...
+}
 
-type MovementSensorConstructor =
-    dyn Fn(ConfigType, Option<BoardType>) -> anyhow::Result<MovementSensorType>;
+impl TryFrom<ResourceName> for ResourceKey {
+    type Error = anyhow::Error;
+    fn try_from(value: ResourceName) -> Result<Self, Self::Error> {
+        let comp_type: &str = &value.subtype;
+        let comp_name = match comp_type {
+            "motor" => crate::common::motor::COMPONENT_NAME,
```
Comment (npmenard) [inline src/common/registry.rs:84]: Is Board missing on purpose?

File: src/common/robot.rs [hunk 2]
```diff
@@ -130,14 +143,18 @@ impl LocalRobot {
             resources: ResourceMap::new(),
         };
 
-        let components = &(config_resp.config.as_ref().unwrap().components);
+        let mut components = config_resp.config.as_ref().unwrap().components.clone();
```
Comment (npmenard) [inline src/common/robot.rs:146]: is there a way we can avoid cloning here? it doesn't look like you are keeping ownership of the cloned components

Comment (npmenard) [reply src/common/robot.rs:146]: I see now why you need it. Do you think we could write the algorithm so that the vector you allocate is a Vec<Option<&ComponentConfig>> this way as you can just take the inner value of the Option and put it back if it needs further processing. Otherwise the Vector will contently move things around in memory which can be problemactic for larger configs. Could also use a Map here. If this is too complicated then I think we should do a Vec<&ComponentConfig> so it takes less space in memory

**Review by npmenard — CHANGES_REQUESTED** [review]
Done with first pass. I don't have major comments but I would like to see some more corner case tests (dependency cylcle, absence of dependencies etc...) If you have time today I'd like a run trough the algorithm to make sure I understood it correctly

File: src/common/registry.rs (hunk 1 above)
Comment (gvaradarajan) [reply src/common/registry.rs:84]: since we always add the board first, I didn't end up needing it. But I can add it

**Review by npmenard — APPROVED** [review]
LGTM
